### PR TITLE
CUDA/HIP: fix tests/test-backend-ops

### LINKED ISSUE #8863: Bug: CUDA/HIP backends assertion crash in test-backend-ops perf
### What happened?

The CUDA and HIP backends crash when running `test-backend-ops perf -o MUL_MAT` with an assertion error
```
llama.cpp/ggml/src/ggml-cuda.cu:1405: GGML_ASSERT(src1->type == GGML_TYPE_F32 || (src1->ne[2] == 1 && src1->ne[3] == 1)) failed
```
Either that assertion is wrong or `ggml_backend_cuda_supports_op` should be updated accordingly.

### Name and Version

version: 3520 (49bf8d47)
built with cc (GCC) 14.2.1 20240802 for x86_64-pc-linux-gnu

### What operating system are you seeing the problem on?

Linux

### Relevant log output

```shell
» build_cuda/bin/test-backend-ops perf -o MUL_MAT
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 3090, compute capability 8.6, VMM: yes
Testing 2 backends

Backend 1/2 (CPU)
  Skipping CPU backend
Backend 2/2 (CUDA0)
  Backend name: CUDA0
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=1,k=256,bs=[1,1],nr=[1,1]):                   8192 runs -    37.49 us/run -       32 kB/run -    0.82 GB/s
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=1,k=256,bs=[10,1],nr=[1,1]):                  8192 runs -   374.99 us/run -      320 kB/run -    0.82 GB/s
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=1,k=256,bs=[10,1],nr=[2,1]):                  8192 runs -   173.72 us/run -      641 kB/run -    3.52 GB/s
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=1,k=256,bs=[10,10],nr=[1,1]):                 8192 runs -   863.99 us/run -     3206 kB/run -    3.54 GB/s
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=1,k=256,bs=[10,10],nr=[2,1]):                 5233 runs -  1728.49 us/run -     6412 kB/run -    3.54 GB/s
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=1,k=256,bs=[10,10],nr=[1,2]): not supported
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=1,k=256,bs=[10,10],nr=[2,2]): not supported
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=16,k=256,bs=[1,1],nr=[1,1]):                  8192 runs -     6.66 us/run -      513 kB/run -   73.48 GB/s
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=16,k=256,bs=[10,1],nr=[1,1]):                 6541 runs -    66.53 us/run -     5130 kB/run -   73.53 GB/s
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=16,k=256,bs=[10,1],nr=[2,1]):                 3271 runs -   133.07 us/run -    10260 kB/run -   73.53 GB/s
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=16,k=256,bs=[10,10],nr=[1,1]):                 655 runs -   665.25 us/run -    51300 kB/run -   73.54 GB/s
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=16,k=256,bs=[10,10],nr=[2,1]):                 328 runs -  1330.71 us/run -   102600 kB/run -   73.53 GB/s
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=16,k=256,bs=[10,10],nr=[1,2]): not supported
  MUL_MAT(type_a=f32,type_b=f32,m=16,n=16,k=256,bs=[10,10],nr=[2,2]): not supported
  MUL_MAT(type_a=f32,type_b=f16,m=16,n=1,k=256,bs=[1,1],nr=[1,1]):                   8192 runs -    12.30 us/run -       24 kB/run -    1.86 GB/s
  MUL_MAT(type_a=f32,type_b=f16,m=16,n=1,k=256,bs=[10,1],nr=[1,1]):             /home/user/llama.cpp/ggml/src/ggml-cuda.cu:1405: GGML_ASSERT(src1->type == GGML_TYPE_F32 || (src1->ne[2] == 1 && src1->ne[3] == 1)) failed
ptrace: Operation not permitted.
No stack.
The program is not being run.
[1]    584842 IOT instruction (core dumped)  build_cuda/bin/test-backend-ops perf -o MUL_MAT
```

### LINKED ISSUE #8864: Bug: HIP backend test-backend-ops perf crashes for FLASH_ATTN_EXT
### What happened?

Running `test-backend-ops perf -o FLASH_ATTN_EXT` fails on the HIP backend with the error
```
Unsupported KV type combination for head_size 64.
By default only f16 KV cache is supported.
Compile with GGML_CUDA_FA_ALL_QUANTS for V cache quantization support.
```

### Name and Version

version: 3520 (49bf8d47)
built with cc (GCC) 14.2.1 20240802 for x86_64-pc-linux-gnu

### What operating system are you seeing the problem on?

Linux

### Relevant log output

```shell
» build_rocm/bin/test-backend-ops perf -o FLASH_ATTN_EXT
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
ggml_cuda_init: found 1 ROCm devices:
  Device 0: AMD Radeon Pro VII, compute capability 9.0, VMM: no
Testing 2 backends

Backend 1/2 (CPU)
  Skipping CPU backend
Backend 2/2 (ROCm0)
  Backend name: ROCm0
  FLASH_ATTN_EXT(hs=64,nh=32,kv=512,nb=1,mask=1,max_bias=0.000000,type_KV=f16):                          8098 runs -   104.30 us/run -     4144 kB/run -   37.89 GB/s
  FLASH_ATTN_EXT(hs=64,nh=32,kv=512,nb=1,mask=1,max_bias=0.000000,type_KV=q8_0): Unsupported KV type combination for head_size 64.
By default only f16 KV cache is supported.
Compile with GGML_CUDA_FA_ALL_QUANTS for V cache quantization support.
                   /home/user/llama.cpp/ggml/src/ggml-cuda/fattn-common.cuh:567: fatal error
ptrace: Operation not permitted.
No stack.
The program is not being run.
[1]    759592 IOT instruction (core dumped)  build_rocm/bin/test-backend-ops perf -o FLASH_ATTN_EXT
```

### PR DESCRIPTION
**Pull request by JohannesGaessler**
Fixes https://github.com/ggerganov/llama.cpp/issues/8864 .
The problem is simply that I forgot to adapt the HIP test logic at some point.

Fixes https://github.com/ggerganov/llama.cpp/issues/8863 .
The problem is that there are test cases that are (according to the corresponding functions) supported with the CUDA backend but not with the CPU backend. As a consequence, when running the tests normally these tests are not executed because there are no CPU results to compare the CUDA results to. In performance mode they are executed though and trigger asserts. Quite frankly though, without a way to assert that the results produced are actually correct such performance numbers would be useless anyways. So I edited the test code in such a way that performance is only evaluated for those ops where correctness can also be tested.

### CODE REVIEW

File: tests/test-backend-ops.cpp [hunk 1]
```diff
@@ -539,7 +539,7 @@ struct test_case {
         return false;
     }
 
-    bool eval_perf(ggml_backend_t backend, const char * op_name) {
+    bool eval_perf(ggml_backend_t backend1, ggml_backend_t backend2, const char * op_name) {
```
Comment (slaren) [inline tests/test-backend-ops.cpp:542]: I don't understand the purpose of this change, the perf mode only uses one backend.

**Comment by slaren** [comment]
> So I edited the test code in such a way that performance is only evaluated for those ops where correctness can also be tested.

This should not be necessary.

**Comment by slaren** [comment]
It seems that the issue is that the CUDA backend does not correctly report in the `supports_op` function what kinds of flash attention operations it can do. The solution is to fix that, not to check also the CPU backend.

**Comment by JohannesGaessler** [comment]
In my opinion the reportedly supported ops being inaccurate and performance tests being executed for cases where the correctness cannot be asserted are both issues. But since the whole point of the tests is to reduce the time developers need to spend on quality assurance I don't want to discuss this at length.

**Review by slaren — APPROVED** [review]

**Pull request merged**